synapse/appservice/__init__.py

```diff
@@ -251,6 +264,15 @@ def get_exlusive_user_regexes(self):
             if regex_obj["exclusive"]
         ]
 
+    def get_groups_for_user(self, user_id):
+        """Get the groups that this user is associated with by this AS
+        """
+        return [
+            regex_obj["group_id"]
+            for regex_obj in self.namespaces[ApplicationService.NS_USERS]
+            if "group_id" in regex_obj and regex_obj["regex"].match(user_id)
+        ]
```

Comment:
I'd probably make this return a generator, which is technically a lot faster as it doesn't allocate. (Just use normal parens rather than square ones)